File: app/models/user_sign.rb
class UserSign < ApplicationRecord
  validates :os_name, inclusion: { in: ['ios', 'android'] }
  # validates_length_of :role_id, is: 16, message: "invalid role id"
  # validates_length_of :sign, in: 116..124, message: "invalid sign"

  # NOTE: when using validates, we can't customize the error message
  # validates :role_id, length: { is: 16 }
  # validates :sign, length: { in: 116..124 }

  validate do
    errors.add(:role_id, 'invalid role id') unless role_id.length == 16
    errors.add(:sign, 'invalid sign') unless sign.length.between?(116,124)
  end

  def as_json options = nil
    super.slice("os_name", "role_id", "role_name", "sign")
  end
end

File: app/javascript/components/UserSign.js
import React from "react"
import PropTypes from "prop-types"

class UserSign extends React.Component {
  render () {
    return (
      <tr>
        <td> {this.props.os_name} </td>
        <td> {this.props.role_name} </td>
        <td> {this.props.role_id} </td>
        <td> {this.props.sign} </td>
      </tr>
    );
  }
}

UserSign.propTypes = {
  os_name: PropTypes.string,
  role_name: PropTypes.string,
  role_id: PropTypes.string,
  sign: PropTypes.string
};
export default UserSign
